nideep/datasets/amfed/amfed.py

```diff
@@ -93,23 +100,150 @@ def find_data_intersection(self):
+            else:
+                if not (os.path.isfile(label_file) and os.path.isfile(frame_file)):
+                    success, msg = self.preprocess_frame(entity, image, count, labels_df=labels_df,
+                                                         landmarks_df=landmarks_df)
+                    record['valid'] = success
+                    record['message'] = msg
+
+            result.append(record)
+            success, image = vidcap.read()
+            count += 1
+
+        return result
+
+    def preprocess_frame(self, entity, frame, frame_number, labels_df=None, landmarks_df=None):
+        label_file = self.__labels_filename(entity, frame_number)
+        frame_file = self.__frame_filename(entity, frame_number)
+
+        time = AMFED.__frame_to_timestamp(frame_number)
+
+        if labels_df is None:
+            labels_df = pd.read_csv(entity.au_label)
+
+        if not landmarks_df is None:
+            landmarks_df = pd.read_csv(entity.landmarks, header=None)
+
+        labels = []
+        aucs = ['Smile', 'AU02', 'AU04', 'Trackerfail']  # TODO remove hardcoded aucs
+
+        label_columns = [e for e in aucs if e in list(labels_df)]
+        if label_columns:
+            label_row = labels_df[labels_df['Time'] <= time].iloc[-1].loc[label_columns]
+            labels = [1 if e >= 50 else 0 for e in label_row]
+
+        if frame_number < landmarks_df.shape[0]:
+            left_eye = (
+                landmarks_df.iloc[[frame_number]][0][frame_number], landmarks_df.iloc[[frame_number]][1][frame_number])
+            right_eye = (
+                landmarks_df.iloc[[frame_number]][2][frame_number], landmarks_df.iloc[[frame_number]][3][frame_number])
+        else:
+            return False, 'Frame number > number of landmark file rows'
+
+        if not all(v == 0 for v in (left_eye + right_eye)) and len(labels) == len(aucs):
+            # print 'Read a new frame from video: ', count
+
+            if labels[-1]:
+                return False, 'Tracker fail: human labeled failure'
+
+            cropped = np.array(crop_face(Image.fromarray(frame), eye_left=left_eye, eye_right=right_eye,
+                                         offset_pct=(0.25, 0.25), dest_sz=(64, 64)))
+
+            # gray_image = cv2.cvtColor(cropped, cv2.COLOR_BGR2GRAY)
+
+            pickle.dump(labels[:-1], open(label_file, 'wb'))
+            cv2.imwrite(frame_file, cropped)  # save frame as JPEG file
+            return True, 'Success'
+        else:
+            return False, 'Tracker fail: no landmarks available'
+
+    def __labels_filename(self, entity, frame_number):
+        video_name = entity.video.split('/')[-1][:-4]
+        return self.cache_dir + video_name + '-labels%d.p' % frame_number
+
+    def __frame_filename(self, entity, frame_number):
+        video_name = entity.video.split('/')[-1][:-4]
+        return self.cache_dir + video_name + '-frame%d.jpg' % frame_number
+
+    @staticmethod
```

Comment:
do not assume a constant and equal frame rate for all videos in this datassets (e.g. FLV and AVI have different frame rates). Instead use Entity.timestamps_sec